File: day03/solve.py
def get_data(test: bool):
    file_name = "test_input.txt" if test else "input.txt"

    with open(file_name, "r") as f:
        return f.read().splitlines()


def _search(bank: str, start: int, end: int, batteries: str) -> str:
    largest = ""
    largest_pos = 0
    for i in range(start, len(bank) - end):
        if bank[i] > largest:
            largest: str = bank[i]
            largest_pos = i

    batteries += largest

    if end == 0:
        # we've reached the end of the digits to find
        return batteries
    else:
        return _search(bank, largest_pos + 1, end - 1, batteries)


def _call_search(data: list[str], num_batteries_to_find: int):
    total = 0

    for bank in data:
        batteries = ""
        total += int(_search(bank, 0, num_batteries_to_find - 1, batteries))

    return total


def part1(data: list[str]):
    return _call_search(data, 2)


def part2(data: list[str]):
    return _call_search(data, 12)


def main():
    data = get_data(test=False)
    print(part1(data))
    print(part2(data))

    # day 3 ez bro
    # I like a nice recursive question


if __name__ == "__main__":
    main()

File: day03/solve.go
package main

import (
	"fmt"
	"log"
	"strconv"
	"strings"

	"github.com/bailey4770/aoc2025/utils"
)

func search(bank, batteries string, start, end int) (string, error) {
	largest := 0
	largestPos := 0
	stop := len(bank) - end
	runes := strings.Split(bank, "")

	for i := start; i < stop; i++ {
		if asInt, err := strconv.Atoi(runes[i]); err != nil {
			return "", err
		} else if asInt > largest {
			largest = asInt
			largestPos = i
		}
	}

	batteries += strconv.Itoa(largest)

	if end == 0 {
		return batteries, nil
	} else {
		return search(bank, batteries, largestPos+1, end-1)
	}
}

func callSearch(data []string, numBatteriesToFind int) (int, error) {
	total := 0

	for _, bank := range data {
		batteries := ""

		res, err := search(bank, batteries, 0, numBatteriesToFind-1)
		if err != nil {
			return 0, nil
		}

		if asInt, err := strconv.Atoi(res); err != nil {
			return 0, err
		} else {
			total += asInt
		}
	}

	return total, nil
}

func part1(data []string) (int, error) {
	numBatteriesToFind := 2
	return callSearch(data, numBatteriesToFind)
}

func part2(data []string) (int, error) {
	numBatteriesToFind := 12
	return callSearch(data, numBatteriesToFind)
}

func main() {
	data, err := utils.LoadString(false, "\n")
	if err != nil {
		log.Fatalf("Error: %v", err)
	}

	res1, err := part1(data)
	if err != nil {
		log.Fatalf("Error: %v", err)
	}
	fmt.Printf("Part 1: %v\n", res1)

	res2, err := part2(data)
	if err != nil {
		log.Fatalf("Error: %v", err)
	}
	fmt.Printf("Part 2: %v\n", res2)
}
